[WIP] Expose API that allows for network resizing images

iskakaushik opened the pull request:
## Description

This PR creates a new image provider `ResizedImage` that allows the user to specify the dimensions of the image that the engine can then use to resize to. This is the engine side PR accounting for the same changes: https://github.com/flutter/engine/pull/8596


## Related Issues

https://github.com/flutter/flutter/issues/26194

## Tests

NEEDS TESTS!!

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

Commit:
Provide a way to resize an image.

Commit:
undo formatting changes

jonahwilliams commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
       throw 'Unable to read data';
-    return await PaintingBinding.instance.instantiateImageCodec(data.buffer.asUint8List());
+    return await PaintingBinding.instance.instantiateImageCodec(
+        data.buffer.asUint8List(), size: size);
```
One line

jonahwilliams commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
+
+  @override
+  Future<_SizeAwareCacheKey> obtainKey(
+      ImageConfiguration configuration) async {
```
The ImageConfiguration already has the size information, we should find a way to use that

jonahwilliams commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -197,6 +214,13 @@ class Image extends StatefulWidget {
+  /// set for resizing to take effect. This parameter indicates to the engine
+  /// that this image must be resized. The image will be sized to the constraints
+  /// of the layout or [width] and [height] regardless of this parameter. This
+  /// parameter is primarily intended for use to reduce the memory usage of
```
```suggestion
  /// parameter is primarily intended to reduce the memory usage of
```

jonahwilliams commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -197,6 +214,13 @@ class Image extends StatefulWidget {
+  /// that this image must be resized. The image will be sized to the constraints
+  /// of the layout or [width] and [height] regardless of this parameter. This
+  /// parameter is primarily intended for use to reduce the memory usage of
+  /// [ImageCache].
```
We should gather some numbers to confirm this is true.

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
+///
+/// * [Image.network] for example usage when `resizeToFit` parameter is set.
+class ResizedImage
+    extends ImageProvider<_SizeAwareCacheKey> {
```
nit: no need for newline here

Hixie commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -215,7 +239,15 @@ class Image extends StatefulWidget {
     Map<String, String> headers,
-  }) : image = NetworkImage(src, scale: scale, headers: headers),
+    bool resizeToFit = false,
+  }) : image = _resizeIfNeeded(
```
why do we do this for network image but not asset image or memory image?

iskakaushik replied:
@Hixie , i wanted to get feedback on this `_resizeIfNeeded` pattern before i go ahead and change the other ones.

Hixie replied:
I don't really understand the pattern as written. It seems very network-specific. I'm having trouble understanding how it generalizes.

goderbauer commented:
@iskakaushik are you still working on this one or should we close it?

iskakaushik commented:
> @iskakaushik are you still working on this one or should we close it?

@goderbauer yes. I am waiting on the engine side to reach a good state first.

Commit:
Merge branch 'master' of github.com:flutter/flutter into expose-resizing-image-api

Commit:
Make it so binding has the target values

Commit:
Revert image provider to master

Commit:
Some more progress

Commit:
Resizing now available for network, memory and file images

Commit:
Add docs for the new targetHeight and targetWidth

Commit:
Add tests and fix test code

jonahwilliams commented:
We should be able to resize any image with this, but FileImage would be another usecase outlined in the original doc:

- Take high resolution picture with phone camera
- Save to filesystem
- Load preview image in flutter app that is much smaller

iskakaushik commented:
@jonahwilliams I added support for memory, file and network images with my latest changes. Also added some tests -- though these only test memory images for now. I will add some more to account for network and memory images.

Commit:
fix more test compile errors

goderbauer commented:
@iskakaushik Do you still have plans for this PR? It is getting a little stale. :)

iskakaushik commented:
closing this for now. will reopen after adding some benchmarks for current image decoding.
